metpy/calc/thermo.py

```diff
@@ -1038,3 +1038,165 @@ def most_unstable_parcel(p, temperature, dewpoint, heights=None,
     theta_e = equivalent_potential_temperature(p_layer, T_layer)
     max_idx = np.argmax(theta_e)
     return p_layer[max_idx], T_layer[max_idx], Td_layer[max_idx]
+
+
+@exporter.export
+@check_units('[temperature]', '[pressure]', '[temperature]')
+def isentropic_interpolation(theta_levels, pressure, temperature, *args, **kwargs):
+    r"""Interpolate data in isobaric coordinates to isentropic coordinates.
+
+    Parameters
+    ----------
+    theta_levels : array
+        One-dimensional array of desired theta surfaces
+    pressure : array
+        One-dimensional array of pressure levels
+    temperature : array
+        Array of temperature
+    args : array, optional
+        Any additional variables will be interpolated to each isentropic level.
+
+    Returns
+    -------
+    isentpr : list
+        List with pressure at each isentropic level, followed by each additional
+        argument interpolated to isentropic coordinates.
+
+    Other Parameters
+    ----------------
+    axis : int, optional
+        The axis corresponding to the vertical in the temperature array, defaults to 0.
+    tmpk_out : bool, optional
+        If true, will calculate temperature and output as the last item in the output list.
+        Defaults to False.
+    max_iters : int, optional
+        The maximum number of iterations to use in calculation, defaults to 50.
+    eps : float, optional
+        The desired absolute error in the calculated value, defaults to 1e-6.
+
+    Notes
+    -----
+    Input variable arrays must have the same number of vertical levels as the pressure levels
+    array. Pressure is calculated on isentropic surfaces by assuming that temperature varies
+    linearly with the natural log of pressure. Linear interpolation is then used in the
+    vertical to find the pressure at each isentropic level. Interpolation method from
+    [Hoerling1993]_. Any additional arguments are assumed to vary linearly with temperature
+    and will be linearly interpolated to the new isentropic levels.
+
+    See Also
+    --------
+    potential_temperature
+
+    """
+    # iteration function to be used later
+    def _isen_iter(pk1, isentlevs2, ka, a, b, pok):
+        ekp = np.exp((-ka) * pk1)
+        t = a * pk1 + b
+        f = isentlevs2 - pok * t * ekp
+        fp = pok * ekp * ((ka) * t - a)
+        return pk1 - (f / fp)
+
+    # Change when Python 2.7 no longer supported
+    # Pull out keyword arguments
+    tmpk_out = kwargs.pop('tmpk_out', False)
+    max_iters = kwargs.pop('max_iters', 50)
+    eps = kwargs.pop('eps', 1e-6)
+    axis = kwargs.pop('axis', 0)
+
+    # Get dimensions in temperature
+    ndim = temperature.ndim
+
+    # Convert units
+    pres = pressure.to('hPa')
+    temperature = temperature.to('kelvin')
+
+    # Make pressure the same shape as temperature
+    for dim in range(ndim):
```

Comment:
I tried that already, array broadcasting later on results in the output becoming the shape of pres and not the shape of temperature.